### PR TITLE
Fix issues in `KeysConverter.ConvertTo` introduced in #8255

### LINKED ISSUE #8456: Error in Menustrip's shortcutkey property
### .NET version

8.0.100-alpha.1.23054.7

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

Yes,it cannot repro on .NET 6.0 & 7.0 SDK builds.

### Issue description

Error in MenuStrip's shortcutkey property, default is "none"

### Steps to reproduce

1. Create a .NET 8.0 WinForms project.
2. Add **MenuStrip** control.
3. Add MenuItem control inside MenuStrip.
4. Click "**Type here**" to set the text property to "TEST".
5. Observe MenuItem's **Shortcutkey** property value in Properties Window.

Or 

1. Extract and open [WinFormsApp28.zip](https://github.com/dotnet/winforms/files/10350587/WinFormsApp28.zip) use VS.
2. Observe MenuItem's Shortcutkey property in Properties Window.

**Actual result:**
Error in MenuStrip's shortcutkey property, displaying "Enter+F12+F11+F10+End+Ctrl+F8+F9+Alt+F4+F5+F6+F7+F1+F2+F3+PgDn+Ins+Home+Del+Shift+PgUp+Back+0+1+2+3+4+5+6+7+8+9+0"
T1
TEST

**Expected result:**
The shortcutkey property in MenuStrip should default to "none".
T2
**More Info:**

1. This issue doesn't repro on .NET Framework.
3. This issue doesn't repro on .NET 7.0, .NET 6.0.
4. VS Version:The latest main VS build.
5. This is a regression issue, it cannot reproduce on .NET version 8.0.100-alpha.1.22601.15. build.
6. Reproduced in version 8.0.100-alpha.1.22607.6 and above.

### PR DESCRIPTION
**Pull request by elachlan**
Also added some tests to better cover the KeysConverter.

Fixes #8456

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8485)

### CODE REVIEW

**Comment by elachlan** [comment]
CC: @merriemcgaw here is the fix for the shortcutkeys.

@Nora-Zhou01 could you please test this to confirm the issue is resolved? Unit tests seem to say so.

**Comment by Nora-Zhou01** [comment]
Tested this PR was fixed that the shortcutkey property in MenuStrip should default to "none".
menu

**Comment by Olina-Zhang** [comment]
@elachlan  In addition, found 2 regression issues. Please take a look.
**Issue1:** Some ShotcutKeys setting causes runtime  unhandled exception, for example: Ctrl+Alt+Shift+A
image
**Issue2**: The runtime display of .Net Core and .net framework are not consistent when setting ShortcutKeys as "Ctrl+Alt+Shift+F1"
image

**Comment by elachlan** [comment]
Okay, for what ever reason the designer has serialized `Ctrl+Alt+Shift+A` as `Keys.A`. This fails because the shortcut isn't valid, it fails the validation logic and throws `InvalidEnumArgumentException`.

I can't see why that is happening.

**Comment by elachlan** [comment]
I've added test cases to prove the the new regressions aren't caused by the `KeysConverter` or `ToolStripMenuItem`. I think the issue outlined above is caused by the serializer or designer.

**Comment by elachlan** [comment]
@Olina-Zhang were the regressions caused by the PR or already present?

**Comment by Olina-Zhang** [comment]
@elachlan **Issue1** and **Issue2** don't reproduce in .Net 7.0 SDK build:
image
image
image
image

And .Net 8.0 without your this PR change has this issue https://github.com/dotnet/winforms/issues/8456 itself and also repro above issue1 with setting ShotcutKeys as Ctrl+Alt+Shift+A, issue2 with setting "Ctrl+Alt+Shift+F1" changes "Enter+F12+F11+F10+End+Ctrl+F8+F9+Alt+F4+F5+F6+F7+F1+F2+F3+PgDn+Ins+Home+Del+Shift+PgUp+Back+0+1+2+3+4+5+6+7+8+9+0" automatically

**Comment by elachlan** [comment]
So this PR has not caused the additional regressions that were found. Which is good to know.

This will require some guidance from the team.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ToolStripMenuItemTests.cs [hunk 1]
```diff
@@ -120,6 +120,26 @@ public void ToolStripMenuItem_GetNativeMenuItemImage_ReturnsExpected(int nativeM
...
 
+        [WinFormsTheory]
+        [InlineData(Keys.None)]
+        [InlineData(Keys.Control | Keys.Alt | Keys.Shift | Keys.A)]
+        [InlineData(Keys.Control | Keys.Alt | Keys.Shift | Keys.F1)]
+        [InlineData(Keys.F1)]
+        public void ToolStripMenuItem_SetShortcutKeys(Keys keys)
+        {
+            using var item = new SubToolStripMenuItem();
+            item.ShortcutKeys=keys;
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ToolStripMenuItemTests.cs:131]:
```suggestion
            item.ShortcutKeys = keys;
```
Same in line 140

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ToolStripMenuItemTests.cs [hunk 2]
```diff
@@ -120,6 +120,26 @@ public void ToolStripMenuItem_GetNativeMenuItemImage_ReturnsExpected(int nativeM
...
+            Assert.Equal(keys, item.ShortcutKeys);
+        }
+
+		[WinFormsTheory]
+        [InlineData(Keys.A)]
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ToolStripMenuItemTests.cs:136]: Consider adding more test cases, e.g. just modifier keys or `Keys.None`.

**Comment by dreddy-work** [comment]
> Okay, for what ever reason the designer has serialized `Ctrl+Alt+Shift+A` as `Keys.A`. This fails because the shortcut isn't valid, it fails the validation logic and throws `InvalidEnumArgumentException`.

@Olina-Zhang , can you open an issue in the designer repo for this. We will investigate this in the designer serialization scope.

**Comment by dreddy-work** [comment]
I will take a look at serialization part later this week and update.

**Review by dmitrii-drobotov — COMMENTED** [review]
ConvertTo fix looks good to me. Regarding missing modifiers issue I can see it reproduces only with 2 or more modifiers, so for example setting "Ctrl+A" works well, but "Ctrl+Shift+A" doesn't.

**Comment by dmitrii-drobotov** [comment]
I think this line also need to be replaced by `if (modifiers.HasFlag(keyValue))` and it should resolve the issue:
https://github.com/dotnet/winforms/blob/1be4ac0524287c704cc14b4f28fb580649bf0f1e/src/System.Windows.Forms/src/System/Windows/Forms/KeysConverter.cs#L243

@elachlan - I assume it would reproduce in a unit test, e.g. check that `converter.ConvertTo(Keys.Control | Keys.Shift | Keys.A, typeof(Enum[]))` converts to `[Keys.Control, Keys.Shift, Keys.A]`.

**Comment by elachlan** [comment]
@Olina-Zhang could your team re-test this please?

**Comment by Nora-Zhou01** [comment]
Tested private binaries built by us for PR: 8485 based on the latest .NET 8.0 SDK build, GH issue 8456 is fixed，and the two regressions caused have been resolved.
**Regressions Issue1:**
R
**Regressions Issue2:**
F1

**Review by dmitrii-drobotov — APPROVED** [review]

**Review by dreddy-work — APPROVED** [review]

**Comment by dreddy-work** [comment]
Thank you @elachlan

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/KeysConverterTests.cs [hunk 3]
```diff
@@ -9,15 +9,52 @@ namespace System.Windows.Forms.Tests
...
+
+        [Theory]
+        [InlineData(Keys.None, "None")]
+        [InlineData(Keys.S, "S")]
+        [InlineData(Keys.Control | Keys.C, "Ctrl+C")]
+        [InlineData(Keys.Control | Keys.Add, "Ctrl+Add")]
+        [InlineData(Keys.Control | Keys.Alt | Keys.D, "Ctrl+Alt+D")]
+        [InlineData(Keys.Control | Keys.Alt | Keys.Shift | Keys.A, "Ctrl+Alt+Shift+A")]
+        [InlineData(Keys.Control | Keys.Alt | Keys.Shift | Keys.F1, "Ctrl+Alt+Shift+F1")]
+        public void ConvertToString_ShouldConvertKeys(Keys keys, string expectedResult)
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/KeysConverterTests.cs:33]: Would `Keys.Alt | Keys.Control | Keys.Shift` still result in "Ctrl+Alt+Shift"?

Comment (elachlan) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/KeysConverterTests.cs:33]: I didn't touch the display order here, but future PRs could cause issues. So it is a good idea for tests. Dmitrii raised this concern in a related PR: https://github.com/dotnet/winforms/pull/8401#discussion_r1071256450

Comment (elachlan) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/KeysConverterTests.cs:33]: On further thought, what I noticed with the keys enum is that the enum value order was always the same after assignment. So the order of the key values doesn't matter. The ordering is set via the `DisplayOrder` dictionary, so it needs to be preserved for the order to also be preserved.

**Pull request merged**